pkg/client/apiutil/dynamicrestmapper.go

```diff
@@ -0,0 +1,216 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiutil
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"time"
+
+	"golang.org/x/time/rate"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+)
+
+// ErrRateLimited is returned by a DynamicRESTMapper method if the number
+// of API calls has exceeded a limit within a certain time period.
+type ErrRateLimited struct {
+	// Duration to wait until the next API call can be made.
+	Delay time.Duration
+}
+
+const errRLMsg = "too many API calls to the DynamicRESTMapper within a timeframe"
+
+func (e ErrRateLimited) Error() string {
+	return fmt.Sprintf("%s (%dns)", errRLMsg, int64(e.Delay))
+}
+
+var errRLRe = regexp.MustCompile(fmt.Sprintf(".*%s \\(([0-9]+)ns\\).*", errRLMsg))
+
+func IsRateLimited(err error) (time.Duration, bool) {
+	if e, ok := err.(ErrRateLimited); ok {
+		return e.Delay, true
+	}
+	if matches := errRLRe.FindStringSubmatch(err.Error()); len(matches) > 1 {
```

Comment:
is there a reason that we really need to do this?  I'd like to switch to xerrors/Go 1.13 errors soon.  Rather that in a follow-up PR than using regexes to parse error messages.

Reply:
I don't think using xerrors here would be too big a change here.

TIL about xerrors.

Reply:
fwiw, i did a quick scan of `client.Client`'s error handling with respect the rest mapper, and it doesn't look like there is any error wrapping going on. so at least for usages in the context of callers receiving errors from `client.Client` methods, the `ErrRateLimited` check should be enough.